### ISSUE TITLE
Investigate simplifying the store and removing the "header cache"

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#2008]
Fast sync to height 50306 with store caches __disabled__ - 

header sync: 111s
body sync: 89s

```
20181122 08:57:37.988 DEBUG grin_servers::common::types - sync_state: sync_status: Initial -> AwaitingPeers(true)
20181122 08:57:45.018 DEBUG grin_servers::common::types - sync_state: sync_status: AwaitingPeers(true) -> HeaderSync { current_height: 0, highest_height: 0 }
20181122 08:57:45.018 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 0, highest_height: 0 } -> BodySync { current_height: 0, highest_height: 0 }
20181122 08:57:45.500 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 0, highest_height: 0 } -> HeaderSync { current_height: 511, highest_height: 50305 }
20181122 08:57:46.119 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 511, highest_height: 50305 } -> HeaderSync { current_height: 1022, highest_height: 50305 }
...
20181122 08:59:06.832 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 50078, highest_height: 50306 } -> HeaderSync { current_height: 50306, highest_height: 50306 }
20181122 08:59:06.832 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 50306, highest_height: 50306 } -> TxHashsetDownload { start_time: 2018-11-22T08:59:06.832808Z, downloaded_size: 0, total_size: 0 }
...
20181122 09:02:17.860 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDownload { start_time: 2018-11-22T09:00:04.313771Z, downloaded_size: 104550209, total_size: 104550209 } -> TxHashsetSetup
...
20181122 09:03:01.049 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetValidation { kernels: 87475, kernel_total: 87481, rproofs: 21261, rproof_total: 127683 } -> TxHashsetSave
20181122 09:03:01.788 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetSave -> TxHashsetDone
20181122 09:03:01.797 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDone -> BodySync { current_height: 0, highest_height: 0 }
20181122 09:03:08.137 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 0, highest_height: 0 } -> HeaderSync { current_height: 50306, highest_height: 50309 }
20181122 09:03:08.755 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 50306, highest_height: 50309 } -> BodySync { current_height: 47425, highest_height: 50309 }
20181122 09:03:10.419 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 47425, highest_height: 50309 } -> BodySync { current_height: 47514, highest_height: 50309 }
...
20181122 09:04:30.677 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 50309, highest_height: 50311 } -> BodySync { current_height: 50310, highest_height: 50311 }
20181122 09:04:30.689 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 50310, highest_height: 50311 } -> NoSync
```

Fast sync (to height 50337) with caches __enabled__ - 

header sync: 80s
body sync: 67s

```
20181122 09:23:30.854 DEBUG grin_servers::common::types - sync_state: sync_status: Initial -> AwaitingPeers(true)
20181122 09:23:37.881 DEBUG grin_servers::common::types - sync_state: sync_status: AwaitingPeers(true) -> HeaderSync { current_height: 0, highest_height: 50331 }
20181122 09:23:38.699 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 0, highest_height: 50331 } -> HeaderSync { current_height: 511, highest_height: 50331 }
...
20181122 09:24:58.322 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 50078, highest_height: 50333 } -> HeaderSync { current_height: 50333, highest_height: 50333 }
20181122 09:24:58.322 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 50333, highest_height: 50333 } -> TxHashsetDownload { start_time: 2018-11-22T09:24:58.322315Z, downloaded_size: 0, total_size: 0 }
...
20181122 09:27:15.551 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDownload { start_time: 2018-11-22T09:25:10.827463Z, downloaded_size: 95351582, total_size: 95351582 } -> TxHashsetSetup
...
20181122 09:27:57.579 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetValidation { kernels: 87504, kernel_total: 87513, rproofs: 21296, rproof_total: 127720 } -> TxHashsetSave
20181122 09:27:58.341 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetSave -> TxHashsetDone
20181122 09:27:58.351 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDone -> BodySync { current_height: 0, highest_height: 0 }
20181122 09:27:59.593 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 0, highest_height: 0 } -> HeaderSync { current_height: 50333, highest_height: 50337 }
...
20181122 09:29:05.088 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 50335, highest_height: 50337 } -> BodySync { current_height: 50336, highest_height: 50337 }
20181122 09:29:05.101 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 50336, highest_height: 50337 } -> NoSync
```

**Comment by antiochp** [comment]
From earlier investigations the largest time spent retrieving headers is in -

```
consensus::next_difficulty()
```

which is used as part of header pow validation.

It needs to iterate back over a window of `60 mins` of headers (approx 60 blocks) as part of the difficulty calculation via the `DifficultyIter`.

tl;dr

Header validation requires us to pull 60 previous headers from the db.
And we do this for each header.

The "header cache" was a naive (but pretty effective) solution to this.
But maybe there are more focused things we can do to solve this without adding cache invalidation problems everywhere else in the system.

**Comment by antiochp** [comment]
#2009 should fix the problems we see in #1936.
But with an increased performance cost - "fast sync" is approx 10% slower without the cache.

Proposing we tackle the perf issues separately.

**Comment by antiochp** [comment]
Experimental branch caching "recent headers" purely for validating header difficulty -

```
20181123 12:01:39.850 DEBUG grin_servers::common::types - sync_state: sync_status: Initial -> AwaitingPeers(true)
20181123 12:01:46.883 DEBUG grin_servers::common::types - sync_state: sync_status: AwaitingPeers(true) -> HeaderSync { current_height: 0, highest_height: 52044 }
20181123 12:01:47.354 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 0, highest_height: 52044 } -> HeaderSync { current_height: 511, highest_height: 52044 }
...
20181123 12:02:54.221 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 52046, highest_height: 52046 } -> TxHashsetDownload { start_time: 2018-11-23T12:02:54.221635Z, downloaded_size: 0, total_size: 0 }
...
20181123 12:04:53.695 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDownload { start_time: 2018-11-23T12:03:20.813047Z, downloaded_size: 100657997, total_size: 100657997 } -> TxHashsetSetup
...
20181123 12:05:38.264 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetValidation { kernels: 94196, kernel_total: 94197, rproofs: 22665, rproof_total: 139473 } -> TxHashsetSave
20181123 12:05:39.042 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetSave -> TxHashsetDone
20181123 12:05:39.053 DEBUG grin_servers::common::types - sync_state: sync_status: TxHashsetDone -> BodySync { current_height: 0, highest_height: 0 }
...
20181123 12:05:45.479 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 0, highest_height: 0 } -> HeaderSync { current_height: 52046, highest_height: 52048 }
...
20181123 12:06:37.356 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 52046, highest_height: 52048 } -> BodySync { current_height: 52047, highest_height: 52048 }
20181123 12:06:37.379 DEBUG grin_servers::common::types - sync_state: sync_status: BodySync { current_height: 52047, highest_height: 52048 } -> NoSync
```
